rascaline/src/calculators/bondatom_neighbor_list.rs

```diff
@@ -0,0 +1,909 @@
+    let dsxdz = dsdz*x2;
+    
+    //let sy = s*y2;  //needed
+    let dsydx = dsdx*y2 + s*dy2dx;
+    let dsydy = dsdy*y2 + s*dy2dy;
+    let dsydz = dsdz*y2;
+    
+    //let t1 = tx*x2 +c;  // needed
+    let dt1dx = dcdx + dtxdx*x2 + tx*dx2dx;
+    let dt1dy = dcdy + dtxdy*x2 + tx*dx2dy;
+    let dt1dz = dcdz + dtxdz*x2;
+    
+    //let t2 = tx*y2;  // needed
+    let dt2dx = dtxdx*y2 + tx*dy2dx;
+    let dt2dy = dtxdy*y2 + tx*dy2dy;
+    let dt2dz = dtxdz*y2;
+    
+    //let t3 = t*y2*y2 +c;  // needed
+    let dt3dx = -dcdx*y2*y2 + 2.*t*y2*dy2dx +dcdx; 
+    let dt3dy = -dcdy*y2*y2 + 2.*t*y2*dy2dy +dcdy;
+    let dt3dz = -dcdz*y2*y2 +dcdz;
+
+    return (
+        // Matrix3::new([
+        //     [tx*x2 +c,  tx*y2,       -sy],
+        //     [tx*y2,     t*y2*y2 + c, sx],
+        //     [sy,       -sx,          c],
+        // ]),
+        Matrix3::new([
+            [dt1dx,  dt2dx, -dsydx],
+            [dt2dx,  dt3dx,  dsxdx],
+            [dsydx, -dsxdx,  dcdx],
+        ]),
+        Matrix3::new([
+            [dt1dy,  dt2dy, -dsydy],
+            [dt2dy,  dt3dy,  dsxdy],
+            [dsydy, -dsxdy,  dcdy],
+        ]),
+        Matrix3::new([
+            [dt1dz,  dt2dz, -dsydz],
+            [dt2dz,  dt3dz,  dsxdz],
+            [dsydz, -dsxdz,  dcdz],
+        ]),
+    );
+}
+
+
+
+/// Manages a list of 'neighbors', where one neighbor is the center of a pair of atoms
+/// (first and second atom), and the other neighbor is a simple atom (third atom).
+/// Both the length of the bond and the distance between neighbors are subjected to a spherical cutoff.
+/// 
+/// Unlike the corresponding pre_calculator, this calculator focuses on storing
+/// the canonical-orientation vector between bond and atom, rather than the bond vector and 'third vector'.
+///
+/// Users can request either a "full" neighbor list (including an entry for both
+/// `i - j` bonds and `j - i` bonds) or save memory/computational by only
+/// working with "half" neighbor list (only including one entry for each `i - j`
+/// bond)
+/// if memory is saved, the order of i and j is that the atom with
+/// the smallest Z (or species ID in general) comes first.
+///
+/// The two first atoms must not be the same atom, but the third atom may be one of them,
+/// if the `bond_conbtribution` option is active
+/// (When periodic boundaries arise, atom which  must not be the same may be images of each other.)
+///
+/// This sample produces a single property (`"distance"`) with three components
+/// (`"vector_direction"`) containing the x, y, and z component of the vector from
+/// the center of the triplet's 'bond' to the triplet's 'third atom', in the bond's canonical orientation.
+/// 
+/// In addition to the atom indexes, the samples also contain a pair and triplet index,
+/// to be able to distinguish between multiple triplets involving the same atoms
+/// (which can occur in periodic boundary conditions when the cutoffs are larger than the unit cell).
```

Comment:
We changed this recently for the usual neighbor list, replacing the pair id with the cell shift for the pair (https://github.com/Luthaf/rascaline/pull/221). Could you do something similar here?

Reply:
sure! I'll look into that

Reply:
do I still need to do that if this is no longer exposed as a calculator?
... it probably needs to be exposed at the precalculator (or system/bond_atom_neighbors) level.

Reply:
you don't need to change the samples, but it might make help to make the rest of the code more understandable. You could also just use a `Vec<BATriplets>` at the pre-calculator level, and remove the need for pair/ba-triplet ids altogether

Reply:
I though the idea of the System cache would be to mainly use TensorBlocks?
But if that's no longer a constraint, I'll switch to only using a Vec<BATriplet> rather than converting back and forth to a tensorblock

Reply:
I'm not sure. For now the code in rascaline accepts everything. There is something similar happening as part of https://github.com/lab-cosmo/metatensor/pull/405, where metatensor is getting a `System` definition to define, export and run MLIP; and there the data is only TensorBlock.

If having `Any` here makes your life easier, I would stick with it, since the goals of the two API are a bit different.

Reply:
ah OK, thanks!